### PR TITLE
Updates dependencies and fixes some issues with IPF

### PR DESCRIPTION
**Pull request by isaac-tfn**
Describe Changes
---
Some versions in requirements were causing errors in the latest release, which this addresses.

In the same PR some issues with IPF around subsets are also addressed.

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `setup.cfg`

### CODE REVIEW

File: src/caf/base/data_structures.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
 from typing import Callable, Literal, Optional, Union
 
 import caf.toolkit as ctk
+from caf.toolkit import translation
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:24]: This has been fixed so shouldn't be needed anymore

File: src/caf/base/data_structures.py [hunk 2]
```diff
@@ -342,8 +343,8 @@ def data(self, value: pd.Series | pd.DataFrame):
             raise TypeError(
                 "data must be a pandas DataFrame or Series. Input " f"value is {value.type}."
             )
-        if isinstance(value, pd.Series):
-            value = value.to_frame()
+        # if isinstance(value, pd.Series):
+        #     value = value.to_frame()
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:347]: Why is this commented out, doesn't `_dataframe_to_dvec` require a dataframe? If it can handle Series please can you update the type annotation for it.

File: src/caf/base/data_structures.py [hunk 3]
```diff
@@ -656,7 +661,7 @@ def translate_zoning(
 
         transposed = self.data.transpose()
         transposed.index.names = [self.zoning_system.column_name]
-        translated = ctk.translation.pandas_vector_zone_translation(
+        translated = translation.pandas_vector_zone_translation(
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:664]: This has been fixed so can be reverted back

File: src/caf/base/data_structures.py [hunk 4]
```diff
@@ -939,7 +944,10 @@ def _generic_dunder(
 
     def __len__(self):
         """Return the length of a DVector, defined as number of cells."""
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:946]: Why are we defining length as number of cells instead of number of rows to be inline with pandas and numpy? If we need number of cells I think this might be better as a size property

File: src/caf/base/data_structures.py [hunk 5]
```diff
@@ -1517,18 +1529,43 @@ def validate_ipf_targets(
...
-                        "Input target DVectors do not have consistent "
-                        f"sums, so ipf will fail target at position {position} doesn't match "
-                        "the first target. It is possible later targets also don't match."
-                    )
+                # Check targets sum to the same, or they can't converge. Potentially could allow
+                # IPF for non-agreeing targets to get as close as possible.
+                if target_sum == 0:
+                    target_sum = target.data.sum()
+                else:
+                    # TODO don't hard code this
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:1562]: What is this TODO referring to?

File: src/caf/base/data_structures.py [hunk 6]
```diff
@@ -2191,26 +2230,59 @@ def check_compatibility(targets: Collection[IpfTarget], adjust: bool = False):
             Whether to change the targets or just report on their compatibility.
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:2230]: Please can you add docs for the new `reference` parameter

File: src/caf/base/data_structures.py [hunk 7]
```diff
@@ -2191,26 +2230,59 @@ def check_compatibility(targets: Collection[IpfTarget], adjust: bool = False):
...
         """
         targ_dict = {i: j for i, j in enumerate(targets)}
+        # add reference to end of targ_dict if exists
+        if reference is not None:
+            targ_dict[len(targets)] = reference
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:2235]: You're using a dict but treating it like a list by just using integer index values for the keys, I don't think it makes a real difference here but is just slightly confusing

File: src/caf/base/data_structures.py [hunk 8]
```diff
@@ -2191,26 +2230,59 @@ def check_compatibility(targets: Collection[IpfTarget], adjust: bool = False):
...
         rmses = {}
         for pos in list(itertools.combinations(reversed(targ_dict), 2)):
-            target_1, target_2 = targ_dict[pos[1]].data, targ_dict[pos[0]].data
-            if target_1.zoning_system != target_2.zoning_system:
+            skip = False
+            target_1, target_2 = targ_dict[pos[1]], targ_dict[pos[0]]
+            # target_2 has subsets, not appropriate for comparison
+            if len(target_2.segmentation.input.subsets) > 0:
                 continue
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:2242]: There are a few times where rmses aren't produced, should we put np.nan or a warning to provide some additional information?

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
Methodology looks good just some comments around docs and fixed caf.toolkit imports and some clarification questions about why / how things work.

Also there were some failed actions.

- Isort needs running to fix pylint issues.
- Pylint is complaining about implicit string concatenation for segmentation:L414 these two strings can be combined into one as they fit on one line.
- Couple of other pylint issues on public methods and positional arguments which can probably be disabled in comments.
